[New Concept Exercise] : Generators (Plane Tickets) (#2729)
exercism/python: exercises/concept/plane-tickets/.docs/introduction.md

# Instructions

A generator in Python is a _callable function_ that returns a [lazy iterator](https://en.wikipedia.org/wiki/Lazy_evaluation).

_Lazy iterators_ are similar to `lists`, and other `iterators`, but with one key difference: They do not store their `values` in memory, but _generate_ their values when needed.

## Constructing a generator

> Constructing a `generator` is a bit different than a normal `function`. You will need the `yield` expression, which we will go into depth with [later](#the-yield-expression).

Reviewer: ```suggestion
Generators are constructed much like other  functions, but require a [`yield` expression](#the-yield-expression), which we will explore  in depth a bit later.
```

Reviewer 2: ```suggestion
Generators are constructed much like other functions, but require a [`yield` expression](#the-yield-expression), which we will explore in depth a bit later.
```

Reviewer 2: The suggestion above just removes some "double space" between words.

> Lets say you want to construct a `generator` that generates all the _squares_ from a list of numbers. You would construct that function like this:
> 
> ```python
> >>> def squares(list_of_numbers):
> >>>     for number in list_of_numbers:
> >>>         yield number ** 2
> ```
> 

Reviewer: ```suggestion
An example is a function that returns the _squares_ from a given list of numbers.  As currently written, all input must be processed before any values can be returned:  

```python
>>> def squares(list_of_numbers):
>>>     squares = []
>>>     for number in list_of_numbers:
>>>         squares.append(number ** 2)
>>>     return squares
```
To convert that function into a _generator_, you would change that function to this:

```python
>>> def squares(list_of_numbers):
>>>     for number in list_of_numbers:
>>>         yield number ** 2
```
```

Reviewer: ..thinking we might want to add a little rationale here for why you'd do this instead of writing a utility function that did one square at a time.  Not sure (yet) what I would add though.

## Using a generator

> It is possible to use any Python _function_ or _object_ that requires an `iterator` as an argument.

Reviewer: ```suggestion
Generators can usually be used in place of most `iterables` in Python. This includes  _functions_ or _objects_ that require an `iterable`/`iterator` as an argument.
```

Reviewer 3: It might also be worth calling out explicitly that there's no `len()`

Author: I added most of this!

> For example, if you want to use the `squares()` generator we just constructed, we simply use:

Reviewer: ```suggestion
For example, the `squares()` generator we just constructed:
```

```python

> >>> list_of_numbers = [1, 2, 3, 4]

Reviewer: ```suggestion
>>> squared_numbers = squares([1, 2, 3, 4])
```

> >>> for square in squares(list_of_numbers):

Reviewer: ```suggestion
>>> for square in squared_numbers:
```

>>>     print(square)
1
4
9
16
```

> You can also get access to the values of a `generator` by using the `next()` function. The `next()` function calls the `__next__()` attribute of a generator.

Reviewer: ```suggestion
Values within a generator can also be produced/accessed via the `next()` function. `next()` calls the `__next__()` attribute of a generator, which "advances" or evaluates the generator code to the `yield` expression, returning the "next" value.
```

```python
square_generator = squares([1, 2])

>>> next(square_generator)
1
>>> next(square_generator)
4
```

> When the `generator` has no more values to return it throws a `StopIteration` error.

Reviewer: ```suggestion
When a `generator` is fully consumed and has no more values to return, it throws a `StopIteration` error.
```

```python
>>> next(square_generator)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```

## The yield expression

> The [yield expression](https://docs.python.org/3.8/reference/expressions.html#yield-expressions) is very similar to the `return` expression. Unlike the `return` expression, `yield` returns a _generator object_ to the caller.
> 
> When `yield` is called, it pauses the execution of the function it is in and returns a value. When `__next__()` is called, it resumes the execution of the function.

Reviewer: ```suggestion
The [yield expression](https://docs.python.org/3.8/reference/expressions.html#yield-expressions) is very similar to the `return` expression. _Unlike_ the `return` expression, `yield` gives up values to the caller at a _specific point_ , suspending evaluation/return of any additional values until they are requested.

When `yield` is evaluated, it pauses the execution of the enclosing function and returns any values of the function _at that point in time_. The function then _stays in scope_, and when `__next__()` is called, execution resumes until `yield` is encountered again.
```

> Note: _`yield` expressions are prohibited to be used outside of functions._

Reviewer: ```suggestion
Note: _Using `yield` expressions is prohibited outside of functions._
```

```python
>>> def infinite_sequence():
>>>     current_number = 0
>>>     while True:
>>>         yield current_number
>>>         current_number += 1

>>> lets_try = yield_expression()
>>> lets_try.__next__()
0
>>> lets_try.__next__()
0
>>> lets_try.__next__()
1

> ```

Reviewer: I like this example, but you can get this same behavior using `range()`, so I think we may want to mix it up a bit to show the power of doing something and then suspending.  How about we use `range()` .. but then do some weird calculation or translation of the numbers that we then `yield`?  

Or we could do a generator version of  [say](https://github.com/exercism/python/blob/main/exercises/practice/say/.docs/instructions.md),  or an [armstrong number](https://github.com/exercism/python/blob/main/exercises/practice/armstrong-numbers/.docs/instructions.md) generator....

## Why generators?

Generators are useful in a lot of applications.

> When working with a large collection, you might not want to put all of that into `memory`. You can use generators to work on data piece-by-piece, this saves memory and improves performance.

Reviewer: ```suggestion
When working with a large collection, you might not want to put all of its values into `memory`. A generator can be used to work on larger data piece-by-piece, saving memory and improving performance.
```

> You can also use it to generate complicated or infinite sequences, like this:

Reviewer: ```suggestion
Generators are also very helpful when a process or calculation is complex, "expensive", or "infinite":
```

> ```python

Reviewer: Same comment as above.  We should consider an example that is complex or expensive or both -- to clearly demonstrate why you'd calculate and then suspend.
